### PR TITLE
Clean up OnceCell

### PR DESCRIPTION
**Pull request by Darksonn**
This cleans up the `OnceCell` a bit.

### CODE REVIEW

**Comment by carllerche** [comment]
@hawkw are you able to take a look at this one? :heart:

**Comment by hawkw** [comment]
> @hawkw are you able to take a look at this one? heart

sure, giving it a look!

File: tokio/src/sync/once_cell.rs [hunk 1]
```diff
@@ -157,179 +231,161 @@ impl<T> OnceCell<T> {
...
-    /// for an ongoing initialization to finish, call
-    /// [`OnceCell::get_or_init`] instead.
+    /// If the `OnceCell` is empty, but some other task is currently trying to
+    /// set the value, this call will fail with [`SetError::InitializingError`].
     ///
     /// [`SetError::AlreadyInitializedError`]: crate::sync::SetError::AlreadyInitializedError
     /// [`SetError::InitializingError`]: crate::sync::SetError::InitializingError
-    /// ['OnceCell::get_or_init`]: crate::sync::OnceCell::get_or_init
     pub fn set(&self, value: T) -> Result<(), SetError<T>> {
         if !self.initialized() {
```
Comment (hawkw) [inline tokio/src/sync/once_cell.rs:260]: take it or leave it: might read a little better if this was 
```rust
if self.initialized() {
    return Err(SetError::AlreadyInitializedError(value));
}

match self.semaphore.try_acquire() {
   // ...
}
```
but, take it or leave it

File: tokio/src/sync/once_cell.rs [hunk 2]
```diff
@@ -157,179 +231,161 @@ impl<T> OnceCell<T> {
...
+    /// this call will wait for that other task to finish, then return the value
+    /// that the other task produced.
+    ///
+    /// If the provided operation is cancelled or panics, the initialization
+    /// attempt is cancelled. If there are other tasks waiting for the value to
+    /// be initialized, one of them will start another attempt at initializing
+    /// the value.
     ///
-    /// This will deadlock if `f` tries to initialize the cell itself.
+    /// This will deadlock if `f` tries to initialize the cell recursively.
```
Comment (hawkw) [inline tokio/src/sync/once_cell.rs:298]: could be nice to have an example here, but not a blocker.

File: tokio/src/sync/once_cell.rs [hunk 3]
```diff
@@ -157,179 +231,161 @@ impl<T> OnceCell<T> {
...
+    /// this call will wait for that other task to finish, then return the value
+    /// that the other task produced.
+    ///
+    /// If the provided operation returns an error, is cancelled or panics, the
+    /// initialization attempt is cancelled. If there are other tasks waiting
+    /// for the value to be initialized, one of them will start another attempt
+    /// at initializing the value.
     ///
-    /// This will deadlock if `f` tries to initialize the cell itself.
+    /// This will deadlock if `f` tries to initialize the cell recursively.
```
Comment (hawkw) [inline tokio/src/sync/once_cell.rs:346]: similarly, an example might be nice.

**Review by hawkw — APPROVED** [review]
this looks great --- the code is much clearer, and the docs are much better IMO. :+1:

File: tokio/src/sync/once_cell.rs (hunk 2 above)
Comment (Darksonn) [reply tokio/src/sync/once_cell.rs:298]: Well, then I would want a normal example too. We can't have the only example be how to do it incorrectly.

**Pull request merged**